app/controllers/WebsiteController.scala

```diff
@@ -0,0 +1,103 @@
+package controllers
+
+import java.util.UUID
+
+import com.mohiva.play.silhouette.api.Silhouette
+import javax.inject._
+import models.{Company, UserRoles, Website, WebsiteCreate, WebsiteKind, WebsiteUpdate, WebsiteUpdateCompany}
+import orchestrators.WebsiteOrchestrator
+import play.api.Logger
+import play.api.libs.json.{JsError, Json}
+import repositories.{CompanyRepository, WebsiteRepository}
+import utils.silhouette.auth.{AuthEnv, WithRole}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class WebsiteController @Inject()(
+  val websiteOrchestrator: WebsiteOrchestrator,
+  val websiteRepository: WebsiteRepository,
+  val companyRepository: CompanyRepository,
+  val silhouette: Silhouette[AuthEnv]
+)(implicit ec: ExecutionContext) extends BaseController {
+
+  val logger: Logger = Logger(this.getClass)
+
+  def fetchWithCompanies() = SecuredAction(WithRole(UserRoles.Admin)).async { implicit request =>
+    for {
+      websites <- websiteRepository.list()
+      websitesWithCompanies <- websiteOrchestrator.toJsonWithCompany(websites)
+    } yield {
+      Ok(Json.toJson(websitesWithCompanies))
+    }
+  }
+
+  def update(uuid: UUID) = SecuredAction(WithRole(UserRoles.Admin)).async(parse.json) { implicit request =>
+    request.body.validate[WebsiteUpdate].fold(
+      errors => Future.successful(BadRequest(JsError.toJson(errors))),
+      websiteUpdate =>
+        for {
+          websiteOption <- websiteRepository.find(uuid)
+          updatedWebsiteOption <- websiteOption.map(website => {
+            websiteRepository.update(websiteUpdate.mergeIn(website)).map(Some(_))
+          }).getOrElse(Future(None))
+          websiteWithCompanyJson <- websiteOrchestrator.toJsonWithCompany(updatedWebsiteOption)
+        } yield {
+          websiteWithCompanyJson.map(Ok(_)).getOrElse(NotFound)
+        }
+    )
+  }
+
+  def updateCompany(uuid: UUID) = SecuredAction(WithRole(UserRoles.Admin)).async(parse.json) { implicit request =>
+    request.body.validate[WebsiteUpdateCompany].fold(
+      errors => Future.successful(BadRequest(JsError.toJson(errors))),
+      websiteUpdate =>
+        for {
+          websiteOption <- websiteRepository.find(uuid)
+          company <- companyRepository.getOrCreate(websiteUpdate.companySiret, Company(
+            siret = websiteUpdate.companySiret,
+            name = websiteUpdate.companyName,
+            address = websiteUpdate.companyAddress,
+            postalCode = websiteUpdate.companyPostalCode,
+            activityCode = websiteUpdate.companyActivityCode,
+          ))
+          updatedWebsite <- websiteOption.map(website =>
+            websiteRepository.update(website.copy(companyId = company.id)).map(Some(_))
+          ).getOrElse(Future(None))
+          websiteWithCompany <- websiteOrchestrator.toJsonWithCompany(updatedWebsite)
+        } yield {
+          websiteWithCompany.map(Ok(_)).getOrElse(InternalServerError)
+        }
+    )
```

Comment:
As discussed, by using a monadic actions like
https://github.com/Kanaka-io/play-monadic-actions
or 
https://github.com/Driox/sorus
or something else, I think there is a lot of this kind of stuff.

we could write something like this:

```
for {
      websiteUpdate <- request.body.validate[WebsiteUpdateCompany] ?| BadRequest(JsError.toJson(errors))
      websiteOption <- websiteRepository.find(uuid) ?| InternalServerError("Cannot find...")
      website <- websiteOption ?| NotFound
      company <- companyRepository.getOrCreate(websiteUpdate.companySiret, Company(
        siret = websiteUpdate.companySiret,
        name = websiteUpdate.companyName,
        address = websiteUpdate.companyAddress,
        postalCode = websiteUpdate.companyPostalCode,
        activityCode = websiteUpdate.companyActivityCode,
      )) ?| InternalServerError("Cannot create...")
      updatedWebsite <- websiteRepository.update(website.copy(companyId = company.id)) ?| InternalServerError("Cannot update...")
      websiteWithCompanyOption <- websiteOrchestrator.toJsonWithCompany(updatedWebsite) ?| InternalServerError("...")
      websiteWithCompany <- websiteWithCompanyOption ?| NotFound("...")
    } yield {
      websiteWithCompany
    }
```

Reply:
Effectivement ça rend la syntaxe plus claire.
Inconvénient mon sens :
- je ne pense pas qu'il faille gérer les cas d'erreurs spécifiques à chaque ligne. Est-ce qu'un classe FutureOpt ou équivalent qui gère simplement le cas du NotFound ne suffirait pas ?
- les libs en question semblent assez peu utilisées